File: server/src/board.ts
import {
  BOARD_SIZE,
  FLEET_ORDER,
  type OrientationName,
  type ShipTypeName,
  isWithinBounds,
  shipCells,
} from "./rules.js";

export type ShotResultType = "Miss" | "Hit" | "Sunk" | "AlreadyShot" | "Invalid";

export interface ShotResult {
  ResultType: ShotResultType;
  Row: number;
  Col: number;
  SunkShipType: ShipTypeName | null;
}

interface Ship {
  Type: ShipTypeName;
  Coordinates: Array<{ Row: number; Col: number }>;
  Hits: Set<string>;
}

function key(r: number, c: number): string {
  return `${r},${c}`;
}

export class Board {
  readonly Ships: Ship[] = [];
  readonly ShotsReceived = new Set<string>();

  get AllShipsPlaced(): boolean {
    return this.Ships.length === FLEET_ORDER.length;
  }

  get AllShipsSunk(): boolean {
    return (
      this.Ships.length > 0 && this.Ships.every((s) => s.Hits.size >= s.Coordinates.length)
    );
  }

  private occupies(row: number, col: number): boolean {
    return this.Ships.some((s) =>
      s.Coordinates.some((c) => c.Row === row && c.Col === col),
    );
  }

  CanPlaceShip(
    type: ShipTypeName,
    row: number,
    col: number,
    orientation: OrientationName,
  ): boolean {
    const coords = shipCells(row, col, type, orientation);
    for (const c of coords) {
      if (!isWithinBounds(c.Row, c.Col) || this.occupies(c.Row, c.Col)) {
        return false;
      }
    }
    return true;
  }

  PlaceShip(
    type: ShipTypeName,
    row: number,
    col: number,
    orientation: OrientationName,
  ): boolean {
    if (!this.CanPlaceShip(type, row, col, orientation)) {
      return false;
    }
    if (this.Ships.some((s) => s.Type === type)) {
      return false;
    }
    const coordinates = shipCells(row, col, type, orientation);
    this.Ships.push({
      Type: type,
      Coordinates: coordinates,
      Hits: new Set(),
    });
    return true;
  }

  FireAt(row: number, col: number): ShotResult {
    if (!isWithinBounds(row, col)) {
      return { ResultType: "Invalid", Row: row, Col: col, SunkShipType: null };
    }
    const k = key(row, col);
    if (this.ShotsReceived.has(k)) {
      return { ResultType: "AlreadyShot", Row: row, Col: col, SunkShipType: null };
    }
    this.ShotsReceived.add(k);

    const hitShip = this.Ships.find((s) =>
      s.Coordinates.some((c) => c.Row === row && c.Col === col),
    );

    if (!hitShip) {
      return { ResultType: "Miss", Row: row, Col: col, SunkShipType: null };
    }
    hitShip.Hits.add(k);
    const sunk = hitShip.Hits.size >= hitShip.Coordinates.length;
    if (sunk) {
      return { ResultType: "Sunk", Row: row, Col: col, SunkShipType: hitShip.Type };
    }
    return { ResultType: "Hit", Row: row, Col: col, SunkShipType: null };
  }
}

File: sentience_takehome/Assets/Scripts/Core/Board.cs
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using UnityEngine;

public class ShotResult
{
    public ShotResultType ResultType;
    public Coordinate Coordinate;
    public ShipType? SunkShipType;

    public ShotResult(ShotResultType resultType, Coordinate coordinate, ShipType? sunkShipType = null)
    {
        ResultType = resultType;
        Coordinate = coordinate;
        SunkShipType = sunkShipType;
    }
}

public class Board
{
    public List<Ship> Ships { get; } = new();
    public HashSet<Coordinate> ShotsReceived { get; } = new();

    public bool AllShipsPlaced => Ships.Count == BattleshipRules.FleetOrder.Length;
    public bool AllShipsSunk => Ships.Count > 0 && Ships.All(ship => ship.IsSunk);
    
    public bool CanPlaceShip(ShipType type, Coordinate start, Orientation orientation)
    {
        var coordinates = BattleshipRules.GetShipCoordinates(start, type, orientation);

        foreach (var coord in coordinates)
        {
            if (!BattleshipRules.IsWithinBounds(coord) || Ships.Any(ship => ship.Occupies(coord)))
            {
                return false;
            }
        }

        return true;
    }

    public bool PlaceShip(ShipType type, Coordinate start, Orientation orientation)
    {
        if (!CanPlaceShip(type, start, orientation))
        {
            return false;
        }
        if (Ships.Any(ship => ship.Type == type))
        {
            return false;
        }

        var coordinates = BattleshipRules.GetShipCoordinates(start, type, orientation);
        Ships.Add(new Ship(type, coordinates));
        return true;
    }

    public ShotResult FireAt(Coordinate coordinate)
    {
        if (!BattleshipRules.IsWithinBounds(coordinate))
        {
            return new ShotResult(ShotResultType.Invalid, coordinate);
        }

        if (ShotsReceived.Contains(coordinate))
        {
            return new ShotResult(ShotResultType.AlreadyShot, coordinate);
        }

        ShotsReceived.Add(coordinate);

        var hitShip = Ships.FirstOrDefault(ship => ship.Occupies(coordinate));

        if (hitShip == null)
        {
            return new ShotResult(ShotResultType.Miss, coordinate);
        }

        hitShip.RegisterHit(coordinate);

        if (hitShip.IsSunk)
        {
            return new ShotResult(ShotResultType.Sunk, coordinate, hitShip.Type);
        }

        return new ShotResult(ShotResultType.Hit, coordinate);
    }

    public bool HasShipAt(Coordinate coordinate)
    {
        return Ships.Any(ship => ship.Occupies(coordinate));
    }

    public bool WasShotAt(Coordinate coordinate)
    {
        return ShotsReceived.Contains(coordinate);
    }

    public List<Coordinate> GetShipCoordinates()
    {
        var coordinates = new List<Coordinate>();

        foreach (var ship in Ships)
        {
            coordinates.AddRange(ship.Coordinates);
        }

        return coordinates;
    }
}
